File: bindings/go/typed_test.go
package kevy

import (
	"testing"
	"time"
)

// The typed surface exercised end to end, including the opinion that a
// protocol error is a Go error here (unlike Cmd, where it is data).
func TestTypedSurface(t *testing.T) {
	db, err := OpenMem()
	if err != nil {
		t.Fatal(err)
	}
	defer db.Close()

	if err := db.Set("k", []byte("v")); err != nil {
		t.Fatal(err)
	}
	v, ok, err := db.Get("k")
	if err != nil || !ok || string(v) != "v" {
		t.Fatalf("Get: %q %v %v", v, ok, err)
	}
	if _, ok, _ := db.Get("absent"); ok {
		t.Fatal("Get(absent) claimed a hit")
	}

	if err := db.SetEx("tmp", []byte("x"), 30*time.Second); err != nil {
		t.Fatal(err)
	}
	ttl, err := db.PTTL("tmp")
	if err != nil || ttl <= 0 || ttl > 30_000 {
		t.Fatalf("PTTL: %d %v", ttl, err)
	}

	if n, _ := db.Incr("hits"); n != 1 {
		t.Fatalf("Incr: %d", n)
	}
	if n, _ := db.IncrBy("hits", 10); n != 11 {
		t.Fatalf("IncrBy: %d", n)
	}

	// Typed layer's opinion: WRONGTYPE surfaces as a Go error.
	if _, err := db.Incr("k"); err == nil {
		t.Fatal("Incr on a plain string should error")
	}

	if got, _ := db.MGet("k", "absent", "tmp"); string(got[0]) != "v" ||
		got[1] != nil || string(got[2]) != "x" {
		t.Fatalf("MGet: %q", got)
	}

	if kind, _ := db.TypeOf("k"); kind != "string" {
		t.Fatalf("TypeOf: %q", kind)
	}
	if ok, _ := db.Expire("k", time.Minute); !ok {
		t.Fatal("Expire claimed miss")
	}
	if keys, _ := db.Keys("*"); len(keys) != 3 {
		t.Fatalf("Keys: %v", keys)
	}
	if n, _ := db.DBSize(); n != 3 {
		t.Fatalf("DBSize: %d", n)
	}
	if n, _ := db.Del("k", "absent"); n != 1 {
		t.Fatalf("Del: %d", n)
	}
	if err := db.FlushAll(); err != nil {
		t.Fatal(err)
	}
	if n, _ := db.DBSize(); n != 0 {
		t.Fatalf("DBSize after FlushAll: %d", n)
	}
}

File: bindings/node/typed.test.js
// The typed-surface smoke for @goliapkg/kevy-node (runs under Bun; the same
// file exercises the Node path once the N-API addon lands).
import { expect, test } from "bun:test";

import { KevyError, open, text } from "./index.js";

// GET on a non-string key: the Bun scalar lane collapses WRONGTYPE to a misuse
// code, so index.js falls back to the framed GET; the typed surface must surface
// the -WRONGTYPE as a thrown KevyError whose taxonomy survived the throw.
test("typed GET on a non-string key throws WRONGTYPE", async () => {
  const db = await open();
  expect(db.cmd("RPUSH", "list", "a")).toBe(1); // key now holds a list
  let caught;
  try {
    db.get("list");
  } catch (e) {
    caught = e;
  }
  expect(caught).toBeInstanceOf(KevyError);
  expect(caught).toBeInstanceOf(Error); // taxonomy + stack both survive the throw
  expect(caught.message).toMatch(/^WRONGTYPE/);
  db.close();
});

test("typed surface", async () => {
  const db = await open(); // in-memory

  db.set("k", "v");
  expect(db.getText("k")).toBe("v");
  expect(db.getText("absent")).toBeUndefined();

  db.set("tmp", "x", { ttlMs: 30_000 });
  const ttl = db.pttl("tmp");
  expect(ttl).toBeGreaterThan(0);
  expect(ttl).toBeLessThanOrEqual(30_000);

  expect(db.incrby("hits", 1)).toBe(1);
  expect(db.incrby("hits", 10)).toBe(11);
  expect(() => db.incrby("k", 1)).toThrow(); // typed layer: WRONGTYPE throws

  const got = db.mget("k", "absent", "tmp");
  expect(text(got[0])).toBe("v");
  expect(got[1]).toBeNull();
  expect(text(got[2])).toBe("x");

  expect(db.expire("k", 60_000)).toBe(true);
  expect(db.keys("*").sort()).toEqual(["hits", "k", "tmp"]);
  expect(db.dbsize()).toBe(3);
  expect(db.del("k", "absent")).toBe(1);
  db.flushall();
  expect(db.dbsize()).toBe(0);

  // cmd() stays neutral: the protocol error is a VALUE here.
  expect(db.cmd("NOSUCHVERB")).toBeInstanceOf(KevyError);

  // callback pub/sub via the pump
  const seen = [];
  db.subscribe("room", (payload, channel) => seen.push([text(payload), channel]));
  expect(db.publish("room", "hi")).toBe(1);
  await new Promise((r) => setTimeout(r, 120)); // > tickMs
  expect(seen).toEqual([["hi", "room"]]);

  db.close();
});
